### PR TITLE
Add design docs section and propose a design

### PR DESCRIPTION
**Pull request by omgjlk**
An attempt at using lore for proposing design documents for review.
Putting up
https://github.com/BonnyCI/projman/wiki/GitHub-Pull-Request-Check-and-Gate-Design
as a first proposed design using this method.

Signed-off-by: Jesse Keating <jkeating@j2solutions.net>

### CODE REVIEW

File: developers/contributing/designs/github-pr-check-gate.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
+This page is a statement of how things should be, not how things currently are. This design document should be used to further work on our project until we can meet the desired design.
+
+# Check
+A check pipeline runs tests on a newly opened or modified pull request. GitHub will send a web hook when a PR is created or modified. Zuul reacts to these hooks by way of a trigger, and a pipeline can be configured to react to such a trigger.
+
+A check pipeline is [http://docs.openstack.org/infra/zuul/zuul.html#pipelines](independent), and tests the change as presented, as if it were merged immediately. The check pipeline will set a commit [status](https://developer.github.com/v3/repos/statuses/) on the commit that is the tip of the PR to `pending`. Zuul names this status the same name as the pipeline that generates it (e.g. check_github).
+
+When the jobs pass or fail, Zuul will set the status to either `success` or `failure`, and will provide a link to the logs URL from the job(s) executions.
+
+**On failure, Zuul will also comment on the pull request indicating the failure.**
```
Comment (ConsoleCatzirl) [inline developers/contributing/designs/github-pr-check-gate.md]: Why is this bold?

File: developers/contributing/designs/github-pr-check-gate.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
+This page is a statement of how things should be, not how things currently are. This design document should be used to further work on our project until we can meet the desired design.
+
+# Check
+A check pipeline runs tests on a newly opened or modified pull request. GitHub will send a web hook when a PR is created or modified. Zuul reacts to these hooks by way of a trigger, and a pipeline can be configured to react to such a trigger.
+
+A check pipeline is [http://docs.openstack.org/infra/zuul/zuul.html#pipelines](independent), and tests the change as presented, as if it were merged immediately. The check pipeline will set a commit [status](https://developer.github.com/v3/repos/statuses/) on the commit that is the tip of the PR to `pending`. Zuul names this status the same name as the pipeline that generates it (e.g. check_github).
```
Comment (ConsoleCatzirl) [inline developers/contributing/designs/github-pr-check-gate.md]: This part needs changing: your link syntax for the independent pipeline is reversed, the url should be in the parentheses.

File: developers/contributing/designs/github-pr-check-gate.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+## Check triggers
+Triggers are via web hooks sent from github. The following hooks are used as triggers:
+* `pr-open`
+* `pr-change`
+* `pr-reopen`
+* `pr-comment` with a comment value of `recheck`
+
+# Gate
+A gate pipeline is a [DependentPipelineManager](http://docs.openstack.org/infra/zuul/zuul.html#pipelines) which creates a queue of changes to be tested together to be merged. Gate pipelines should be triggered when a pull request has reached a state in which it _could_ be merged and humans have approved the change to be merged. A gate pipeline will execute a series of tests which may or may not be the same tests as the check pipeline, and it will do so in an efficient way regarding other changes in the same queue. Gate pipelines are triggered through GitHub events as well.
```
Comment (ConsoleCatzirl) [inline developers/contributing/designs/github-pr-check-gate.md]: nit: I'm not sure about the use of "could be merged" here, I would prefer replacing that with either "should be merged" or "is ready to be merged".

File: developers/contributing/designs/github-pr-check-gate.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+**A gate pipeline has requirements before it can be entered. For GitHub, we require that a change has passed the check pipeline already, and that a human with write access has approved the change for merging, and that no other humans with write access have requested changes via a review.**
+
+When the jobs attached to the pipeline start, a status named for the pipeline will be set to `pending`.
+
+When the jobs attached to the pipeline pass, the status will be set to `success`, and the change will be merged via the GitHub API.
+
+If the jobs attached to the pipeline fail, the status will be set to `failure`.
+
+**On failure, Zuul will also comment on the pull request indicating the failure.**
```
Comment (ConsoleCatzirl) [inline developers/contributing/designs/github-pr-check-gate.md]: Why is this bold?

File: developers/contributing/designs/github-pr-check-gate.md [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+## Gate triggers
+Triggers are via web hooks sent from github. The following hooks are used as triggers:
+* **`status` from BonnyCI on HEAD of PR with a state of `success`**
```
Comment (ConsoleCatzirl) [inline developers/contributing/designs/github-pr-check-gate.md]: This should mention that the status comes from the check pipeline.

File: developers/contributing/designs/github-pr-check-gate.md [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+* **`pull_request_review` with an event of `approve`**
+* `pr-comment` with a comment value of `gate-recheck`
+
+## Gate requirements
+Requirements are based on the state of the pull request, which will need to be examined to determine.
+* **[reviews](https://developer.github.com/v3/pulls/reviews/#list-reviews-on-a-pull-request) with state of `approve` from at at least one user with write access.**
+* **[reviews](https://developer.github.com/v3/pulls/reviews/#list-reviews-on-a-pull-request) absence of any reviews with state `REQUEST_CHANGES` from users with write access**
+* **`status` from BonnyCI on HEAD of PR with a state of `success`**
```
Comment (ConsoleCatzirl) [inline developers/contributing/designs/github-pr-check-gate.md]: This should also mention the check pipeline.

**Review by ConsoleCatzirl — CHANGES_REQUESTED** [review]

File: developers/contributing/designs/github-pr-check-gate.md (hunk 2 above)
Comment (omgjlk) [reply developers/contributing/designs/github-pr-check-gate.md]: Fixed.

File: developers/contributing/designs/github-pr-check-gate.md (hunk 3 above)
Comment (omgjlk) [reply developers/contributing/designs/github-pr-check-gate.md]: Fixed.

File: developers/contributing/designs/github-pr-check-gate.md (hunk 5 above)
Comment (omgjlk) [reply developers/contributing/designs/github-pr-check-gate.md]: Fixed

File: developers/contributing/designs/github-pr-check-gate.md (hunk 6 above)
Comment (omgjlk) [reply developers/contributing/designs/github-pr-check-gate.md]: Fixed.

**Review by missaugustina — CHANGES_REQUESTED** [review]
Right now this isn't very clear and seems pretty disorganized. I'm working on some edits that I'll propose shortly.

**Comment by missaugustina** [comment]
I added my edit here - https://github.com/BonnyCI/lore/blob/7e569382920d5aae2ee0859991ebb7f6f86bec86/developers/contributing/designs/github-pr-check-gate.md

**Review by missaugustina — APPROVED** [review]

File: developers/contributing/README.md [hunk 7]
```diff
@@ -38,3 +39,10 @@ Anyone may review or comment on pull requests, but a member of the BonnyCI Capta
 
 
 If a pull request is a work in progress, please start the title with `[WIP]`. It is not necessary to include `[WIP]` in commit messages, but if you do, please remove it before it gets merged.
+
+
+## Design Documents
+Changes to BonnyCI should follow a design document. To propose a design document, open a pull requests to the lore repository. BonnyCI contributors will review and debate the proposed document via pull request comments until the document is accepted. When the pull request is merged, development can begin.
```
Comment (mlangbehn) [inline developers/contributing/README.md]: Should the reference to the lore repo be a link?

**Review by mlangbehn — COMMENTED** [review]
Just a quick inline question.

File: developers/contributing/README.md (hunk 7 above)
Comment (omgjlk) [reply developers/contributing/README.md]: Made it a link.

**Review by ConsoleCatzirl — APPROVED** [review]

**Pull request merged**